spec/shared_stripe_examples/product_examples.rb

```diff
@@ -0,0 +1,137 @@
+    expect(data[product2.id]).to_not be_nil
+    expect(data[product2.id][:id]).to eq("prod_456")
+    expect(data[product2.id][:name]).to eq("My Other Product")
+  end
+
+  it "retrieves a stripe product" do
+    original = stripe_helper.create_product(idless_attributes)
+    product = Stripe::Product.retrieve(original.id)
+
+    expect(product.id).to eq(original.id)
+    expect(product.name).to eq(original.name)
+  end
+
+  it "updates a stripe product" do
+    stripe_helper.create_product(id: "prod_XYZ", name: "Future Product")
+
+    product = Stripe::Product.retrieve("prod_XYZ")
+    expect(product.name).to eq("Future Product")
+
+    product.name = "Updated Product"
+    product.save
+    product = Stripe::Product.retrieve("prod_XYZ")
+    expect(product.name).to eq("Updated Product")
+  end
+
+  it "cannot retrieve a stripe product that doesn't exist" do
+    expect { Stripe::Product.retrieve('nope') }.to raise_error {|e|
+      expect(e).to be_a Stripe::InvalidRequestError
+      expect(e.param).to eq("product")
+      expect(e.http_status).to eq(404)
+    }
+  end
+
+  it "deletes a stripe product" do
+    stripe_helper.create_product(id: "prod_DEL", name: "Aging Product")
+
+    product = Stripe::Product.retrieve("prod_DEL")
+    expect(product).to_not be_nil
+
+    product.delete
+
+    expect { Stripe::Product.retrieve("prod_DEL") }.to raise_error {|e|
+      expect(e).to be_a Stripe::InvalidRequestError
+      expect(e.param).to eq("product")
+      expect(e.http_status).to eq(404)
+    }
+  end
+
+  it "retrieves all products" do
+    stripe_helper.create_product(id: "prod_123", name: "First Product")
+    stripe_helper.create_product(id: "prod_456", name: "Second Product")
+
+    all = Stripe::Product.all
+    expect(all.count).to eq(2)
+    expect(all.map &:id).to include("prod_123", "prod_456")
+    expect(all.map &:name).to include("First Product", "Second Product")
+  end
+
+  it 'retrieves products with limit' do
+    101.times do |i|
+      stripe_helper.create_product(id: "Product #{i}", name: "My Product ##{i}")
+    end
+    all = Stripe::Product.all(limit: 100)
+
+    expect(all.count).to eq(100)
+  end
+
+  describe "Validation", :live => true do
+    let(:params) { stripe_helper.create_product_params }
+    let(:subject) { Stripe::Product.create(params) }
+
+    describe "Required Parameters" do
+      after do
+        params.delete(@attribute_name)
+        message = "Missing required param: #{@attribute_name}."
+        expect { subject }.to raise_error(Stripe::InvalidRequestError, message)
+      end
+
+      it("requires a name") { @attribute_name = :name }
+      it("requires a type") { @attribute_name = :type }
+    end
+
+    #describe "Inclusion" do
+    #  it "validates inclusion of type in 'good' or 'service'" do
+    #    expect {
+    #      Stripe::Product.create(params.merge({type: "OOPS"}))
+    #    }.to raise_error(Stripe::InvalidRequestError, "Invalid type: must be one of good or service")
```

Comment:
For some reason, this test was failing for me with the following attempted implementation: https://github.com/s2t2/stripe-ruby-mock/pull/2/files#diff-dca0e699db8ef6e9c62c24f836fdb3d3R37